File: 2573-remove-nodes-from-linked-list/2573-remove-nodes-from-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNodes(ListNode head) {
        Stack<ListNode> stack = new Stack<>();
        ListNode curr = head;

        while(curr != null){
            while(!stack.isEmpty() && stack.peek().val < curr.val){
                stack.pop();
            }

            stack.push(curr);
            curr = curr.next;
        }

        ListNode newHead = null;
        if(!stack.isEmpty()){
            newHead = stack.firstElement();
        }

        for(int i = 0; i < stack.size() -1; i++){
            stack.get(i).next = stack.get(i+1);
        }
        stack.peek().next = null;

        return newHead;
    }
}

File: 2573-remove-nodes-from-linked-list/2573-remove-nodes-from-linked-list.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def removeNodes(self, head):
        """
        :type head: Optional[ListNode]
        :rtype: Optional[ListNode]
        """
        stack = []
        cur = head

        while cur:

            while stack and stack[-1].val < cur.val:
                stack.pop()
            
            stack.append(cur)
            cur = cur.next
        
        for i in range(len(stack) - 1):
            stack[i].next = stack[i + 1]
            
        stack[-1].next = None

        return stack[0] if stack else None